go/vt/sidecardb/identifier_cache.go

```diff
@@ -0,0 +1,119 @@
+/*
+Copyright 2023 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sidecardb
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	vtrpcpb "vitess.io/vitess/go/vt/proto/vtrpc"
+	"vitess.io/vitess/go/vt/sqlparser"
+	"vitess.io/vitess/go/vt/vterrors"
+)
+
+// IdentifierCache provides a read through cache of sidecar database
+// identifiers; loading the values from an opaque backend database
+// using a provided load function.
+type IdentifierCache struct {
+	// Lazily loaded cache of sidecar database identifiers by keyspace.
+	// The key is a keyspace name string and the val is an sqlparser
+	// string built from an IdentifierCS using the sidecar database
+	// name stored in the backend database read in the provided load
+	// function.
+	sidecarDBIdentifiers sync.Map
+
+	// The callback used to load the values from the database into
+	// the cache.
+	load func(context.Context, string) (string, error)
+}
+
+const (
+	ErrIdentifierCacheUninitialized  = "sidecar database identifier cache is not initialized"
+	ErrIdentifierCacheNoLoadFunction = "the load from database function has not been set"
+	identifierCacheLoadTimeout       = 30 * time.Second
+)
+
+var identifierCache atomic.Value // *IdentifierCache singleton
+
+// NewIdentifierCache returns an initialized cache. This is a
+// singleton so if you call New multiple times you will get the
+// same instance. If the cache has already been initialized then
+// it will return false indicating that your New call did not
+// create a new instance.
+func NewIdentifierCache(loadFunc func(context.Context, string) (string, error)) (*IdentifierCache, bool) {
+	created := identifierCache.CompareAndSwap(nil, &IdentifierCache{
+		load:                 loadFunc,
+		sidecarDBIdentifiers: sync.Map{},
+	})
+	return identifierCache.Load().(*IdentifierCache), created
+}
+
+func GetIdentifierCache() (*IdentifierCache, error) {
+	if identifierCache.Load() == nil {
+		return nil, vterrors.New(vtrpcpb.Code_INTERNAL, ErrIdentifierCacheUninitialized)
+	} else {
+		return identifierCache.Load().(*IdentifierCache), nil
+	}
+}
```

Comment:
It would be easier at other places if we can merge these 2 methods.

Reply:
Merge what two methods? GetIdentifierCache() and Get()? I don't think I like that, although it would be easier for the caller in some cases (specifically where all you want to do is Get a single value) -- in general I think it makes it a little awkward. I think you're right though in that there's no reason NOT to add a convenience function for those specific cases:
```
// GetIdentifierForKeyspace is a convenience function that returns
// the sidecar database identifier for the provided keyspace.
func GetIdentifierForKeyspace(keyspace string) (string, error) {
	cache, err := GetIdentifierCache()
	if err != nil {
		return "", err
	}
	return cache.Get(keyspace)
}
```

I will do that.

Reply:
My thinking was when someone calls into `GetIdentifierCache` then `IdentifierCache` gets created if not already done.

Reply:
This way we do not have to take care of the places where they have to be initialized first before using it.

Reply:
This doesn't work, as the caller needs to provide the load function (which in practice usually means they need to have a topo server handy).

Reply:
The way it is now, the user does not need to know anything about the implementation used (only the creator does). I like this abstraction and IMO it simplifies things for the user.